lib/Dialect/FIRRTL/Transforms/LowerSignatures.cpp

```diff
@@ -0,0 +1,416 @@
+        assert(groundType.isGround() && "only ground types are expected here");
+        newPorts.push_back(
+            {{StringAttr::get(ctx, name), groundType,
+              isFlip ? Direction::Out : Direction::In,
+              symbolsForFieldIDRange(port.sym, fieldID, fieldID), port.loc,
+              annosForFieldIDRange(port.annotations, fieldID, fieldID)},
+             portID,
+             newPorts.size(),
+             fieldID});
+        return success();
+      });
+}
+
+// compute a new moduletype from an old module type and lowering convention.
+// Also compute a fieldID map from port, fieldID -> port
+static LogicalResult computeLowering(FModuleLike mod, Convention conv,
+                                     PortConversion &newPorts) {
+  for (auto [idx, port] : llvm::enumerate(mod.getPorts()))
+    if (computeLoweringImpl(
+            mod, newPorts, conv, idx, port, port.direction == Direction::Out,
+            port.name.getValue(), type_cast<FIRRTLType>(port.type), 0)
+            .failed())
+      return failure();
+  return success();
+}
+
+static LogicalResult lowerModuleSignature(FModuleLike module, Convention conv,
+                                          AttrCache &cache,
+                                          PortConversion &newPorts) {
+  ImplicitLocOpBuilder theBuilder(module.getLoc(), module.getContext());
+  if (computeLowering(module, conv, newPorts).failed())
+    return failure();
+  if (auto mod = dyn_cast<FModuleOp>(module.getOperation())) {
+    Block *body = mod.getBodyBlock();
+    theBuilder.setInsertionPointToStart(body);
+    auto oldNumArgs = body->getNumArguments();
+
+    // Compute the replacement value for old arguments
+    // This creates all the new arguments and produces bounce wires when
+    // necessary
+    SmallVector<Value> bounceWires(oldNumArgs);
+    for (auto &p : newPorts) {
+      auto newArg = body->addArgument(p.type, p.loc);
+      // Get or create a bounce wire for changed ports
+      // For unmodified ports, move the uses to the replacement port
+      if (p.fieldID != 0) {
+        auto &wire = bounceWires[p.portID];
+        if (!wire)
+          wire = theBuilder
+                     .create<WireOp>(module.getPortType(p.portID),
+                                     module.getPortNameAttr(p.portID),
+                                     NameKindEnum::InterestingName)
+                     .getResult();
+      } else {
+        bounceWires[p.portID] = newArg;
+      }
+    }
+    // replace old arguments.  Somethings get dropped completely, like
+    // zero-length vectors.
+    for (auto idx = 0U; idx < oldNumArgs; ++idx) {
+      if (!bounceWires[idx]) {
+        bounceWires[idx] =
+            theBuilder.create<WireOp>(module.getPortType(idx)).getResult();
+      }
+      body->getArgument(idx).replaceAllUsesWith(bounceWires[idx]);
+    }
+
+    // Goodby old ports, now ResultID in the PortInfo is correct.
```

Comment:
```suggestion
    // Goodbye old ports, now ResultID in the PortInfo is correct.
```

Reply:
Ouch.  That's quite the hard-hitting review!